WalletWasabi/Blockchain/TransactionOutputs/CoinsRegistry.cs

```diff
@@ -393,17 +394,50 @@ public ICoinsView AsAllCoinsView()
 
 	public ICoinsView Available() => AsCoinsView().Available();
 
-	public ICoinsView ChildrenOf(SmartCoin coin) => AsCoinsView().ChildrenOf(coin);
-
 	public ICoinsView CoinJoinInProcess() => AsCoinsView().CoinJoinInProcess();
 
 	public ICoinsView Confirmed() => AsCoinsView().Confirmed();
 
-	public ICoinsView DescendantOf(SmartCoin coin) => AsCoinsView().DescendantOf(coin);
+	public ImmutableArray<SmartCoin> DescendantOfAndSelf(SmartCoin coin)
+	{
+		lock (Lock)
+		{
+			return DescendantOfAndSelfNoLock(coin);
+		}
+	}
+
+	/// <summary>Gets descendant coins of the given coin - i.e. all coins that spent the input coin, all coins that spent those coins, etc.</summary>
+	/// <remarks>Callers must acquire <see cref="Lock"/> before calling this method.</remarks>
+	private ImmutableArray<SmartCoin> DescendantOfAndSelfNoLock(SmartCoin coin)
+	{
+		HashSet<SmartCoin> allCoins = new(Coins);
+		allCoins.UnionWith(SpentCoins);
+
+		IEnumerable<SmartCoin> Generator(SmartCoin parentCoin, bool addSelf)
+		{
+			IEnumerable<SmartCoin> childrenOf = parentCoin.SpenderTransaction is not null
+				? allCoins.Where(x => x.TransactionId == parentCoin.SpenderTransaction.GetHash()) // Inefficient.
```

Comment:
This would be much cleaner and more efficient with #11516 's factoring. Anyway.

Reply:
> Could we just add the dictionary you had in that PR? As a small atomic PR?

I will look into it, the smallest PR possible with the dictionary. This structure makes a lot of sense because the values are by tx anyway.